sys/amd64/amd64/pmap.c

```diff
@@ -12311,4 +12312,70 @@ DB_SHOW_COMMAND(ptpages, pmap_ptpages)
 		    (vm_offset_t)pmap->pm_pmltop)), NUP4ML4E, PG_V);
 	}
 }
+
+/***********************
+ * zcond functionality *
+************************/
+struct pmap zcond_pmap;
+vm_offset_t zcond_patch_va;
+
+static void
+pmap_zcond_init(const void *unused) {
+    vm_offset_t kern_start, kern_end;
+    vm_page_t dummy_page;
+
+    kern_start = virtual_avail;
+    kern_end = kernel_vm_end;
+
+	memset(&zcond_pmap, 0, sizeof(zcond_pmap));
+	PMAP_LOCK_INIT(&zcond_pmap);
+	pmap_pinit(&zcond_pmap);
+	pmap_copy(&zcond_pmap, kernel_pmap, kern_start,
+	    kern_end - kern_start, kern_start);
+
+    zcond_patch_va = kva_alloc(PAGE_SIZE);
+    dummy_page = vm_page_alloc_noobj(VM_ALLOC_WIRED);
+    pmap_enter(&zcond_pmap, zcond_patch_va, dummy_page, VM_PROT_WRITE, PMAP_ENTER_WIRED, 0);
+    kva_free(zcond_patch_va, PAGE_SIZE);
+}
+SYSINIT(zcond_pmap, SI_SUB_ZCOND, SI_ORDER_SECOND, pmap_zcond_init, NULL);
+
+void
+pmap_qenter_zcond(vm_page_t m) {
+    pt_entry_t oldpte, pa;
+    pt_entry_t *pte;
+    int cache_bits;
+
+    oldpte = 0;
+    pte = pmap_pte(&zcond_pmap, zcond_patch_va);
+
+    cache_bits = pmap_cache_bits(&zcond_pmap, m->md.pat_mode, false);
+    pa = VM_PAGE_TO_PHYS(m) | cache_bits;
+    if ((*pte & (PG_FRAME | X86_PG_PTE_CACHE)) != pa) {
+            oldpte |= *pte;
+            pte_store(pte, pa | pg_g | pg_nx | X86_PG_A |
+                X86_PG_M | X86_PG_RW | X86_PG_V);
+    }
+
+    if (__predict_false((oldpte & X86_PG_V) != 0))
+            pmap_invalidate_range(&zcond_pmap, zcond_patch_va, zcond_patch_va + PAGE_SIZE);
```

Comment:
```suggestion
    invplg(zcond_patch_va);
```
Invalidate range će potencijalno okinut novi IPI, što će panicat.